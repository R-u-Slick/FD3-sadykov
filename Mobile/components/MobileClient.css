.MobileClient
{
  border-bottom: 1px solid gray;
}

.MobileClientStatusblocked {
  background-color: red;
}

.MobileClientStatusactive {
  background-color: green;
}

th {
  text-align: left;
  width: 120px
}

