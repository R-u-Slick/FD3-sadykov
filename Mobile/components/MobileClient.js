import React from 'react';
import PropTypes from 'prop-types';
import {clientEvents} from './events';
import {formEvents} from './events';

import './MobileClient.css';

class MobileClient extends React.PureComponent {

  static propTypes = {
    info:PropTypes.shape({
      id: PropTypes.number.isRequired,
      surname: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      fathersName: PropTypes.string.isRequired,
      balance: PropTypes.number.isRequired,
    }),
    mode: PropTypes.number.isRequired,
  };

  state = {
    formChanged: false,
  };

  componentDidMount = () => {
    formEvents.addListener('formChanged', this.formChanged);
    formEvents.addListener('formChangedCancel',this.formChangedCancel);
    formEvents.addListener('formCancelled', this.formChangedCancel);
  };

  componentWillUnmount = () => {
    formEvents.removeListener('formChanged', this.formChanged);
    formEvents.removeListener('formChangedCancel',this.formChangedCancel);
    formEvents.removeListener('formCancelled', this.formChangedCancel);
  };

  formChanged = (changed) => {
    this.setState({formChanged: changed});
  }

  formChangedCancel = () => {
    this.setState({formChanged: false});
  }

  clientDelete = () => {
    clientEvents.emit('clientDeleted',this.props.info.id);
  }

  clientEdit = () => {
    clientEvents.emit('clientEdited', this.props.info.id);
  }

  render() {
    console.log("MobileClient id="+this.props.info.id+" render");
    let status;
    if (this.props.info.balance>=0) {
      status='active';
    }
    else {
      status='blocked';
    } 
    return (
    <tr className='MobileClient'>
      <td className='MobileClientBalance'>{this.props.info.surname}</td>
      <td className='MobileClientFIO'>{this.props.info.name}</td>
      <td className='MobileClientBalance'>{this.props.info.fathersName}</td>
      <td className='MobileClientFIO'>{this.props.info.balance}</td>
      <td className={'MobileClientStatus'+status}>{status}</td>
      <td className='MobileClientEditButton'><input className='edit-button' type='button' value='Редактировать' onClick={this.clientEdit} disabled={this.state.formChanged}/></td>
      <td className='MobileClientDeleteButton'><input className='edit-button' type='button' value='Удалить' onClick={this.clientDelete} disabled={this.props.mode}/></td>
    </tr>
    )    
  }
}

export default MobileClient;
