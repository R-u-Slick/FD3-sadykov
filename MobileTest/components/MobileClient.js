import React from 'react';
import PropTypes from 'prop-types';
import {clientEvents} from './events';
import {formEvents} from './events';

import './MobileClient.css';

class MobileClient extends React.PureComponent {

  static propTypes = {
    info:PropTypes.shape({
      id: PropTypes.number.isRequired,
      surname: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      fathersName: PropTypes.string.isRequired,
      balance: PropTypes.number.isRequired,
    }),
  };

  state = {
  };

  clientDelete = () => {
    clientEvents.emit('clientDeleted',this.props.info.id);
  }

  clientEdit = () => {
    clientEvents.emit('clientEdited', this.props.info.id);
  }

  render() {
    console.log("MobileClient id="+this.props.info.id+" render");
    let status;
    if (this.props.info.balance>=0) {
      status='active';
    }
    else {
      status='blocked';
    } 
    return (
    <tr className='MobileClient'>
      <td className='MobileClientBalance'>{this.props.info.surname}</td>
      <td className='MobileClientFIO'>{this.props.info.name}</td>
      <td className='MobileClientBalance'>{this.props.info.fathersName}</td>
      <td className='MobileClientFIO'>{this.props.info.balance}</td>
      <td className={'MobileClientStatus'+status}>{status}</td>
      <td className='MobileClientEditButton'><input className='edit-button' type='button' value='Редактировать' onClick={this.clientEdit} disabled={this.state.formChanged}/></td>
      <td className='MobileClientDeleteButton'><input className='delete-button' type='button' value='Удалить' onClick={this.clientDelete}/></td>
    </tr>
    )    
  }
}

export default MobileClient;
