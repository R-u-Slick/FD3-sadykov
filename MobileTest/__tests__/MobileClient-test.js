"use strict";

import React from 'react';
import renderer from 'react-test-renderer';

import MobileClient from '../components/MobileClient';

test('работа кнопки "Редактировать"', () => {
  // создаём тестовую версию компонента
  let client={id:101, surname:"Иванов", name:"Иван", fathersName:"Иванович", balance:200};
  const component = renderer.create(
  <MobileClient key={client.id} info={client}/>
  );
  // получаем снэпшот (HTML-снимок) компонента для сверки, что вёрстка не испортилась
  let componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();
  // найдём в вёрстке компонента кнопку редактирования клиента
  const buttonElem = component.root.find( el => el.props.className=='edit-button'); 
  // и "нажмём" на неё
  buttonElem.props.onClick();
  // получаем изменённый снэпшот
  componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();
   // "нажмём" кнопку ещё раз
   buttonElem.props.onClick();
     // и получаем окончательный снэпшот
   componentTree=component.toJSON();
   expect(componentTree).toMatchSnapshot();
  }
);

test('работа кнопки "Удалить"', () => {
  // создаём тестовую версию компонента
  let client={id:101, surname:"Иванов", name:"Иван", fathersName:"Иванович", balance:200};
  const component = renderer.create(
  <MobileClient key={client.id} info={client}/>
  );
  // получаем снэпшот (HTML-снимок) компонента для сверки, что вёрстка не испортилась
  let componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();
  // найдём в вёрстке компонента кнопку удаления клиента
  const buttonElem = component.root.find( el => el.props.className=='delete-button'); 
  // и "нажмём" на неё
  buttonElem.props.onClick();
  // получаем изменённый снэпшот
  componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();
  }
);

