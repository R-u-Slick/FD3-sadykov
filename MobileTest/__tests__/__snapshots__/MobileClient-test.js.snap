// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`работа кнопки "Редактировать" 1`] = `
<tr
  className="MobileClient"
>
  <td
    className="MobileClientBalance"
  >
    Иванов
  </td>
  <td
    className="MobileClientFIO"
  >
    Иван
  </td>
  <td
    className="MobileClientBalance"
  >
    Иванович
  </td>
  <td
    className="MobileClientFIO"
  >
    200
  </td>
  <td
    className="MobileClientStatusactive"
  >
    active
  </td>
  <td
    className="MobileClientEditButton"
  >
    <input
      className="edit-button"
      disabled={undefined}
      onClick={[Function]}
      type="button"
      value="Редактировать"
    />
  </td>
  <td
    className="MobileClientDeleteButton"
  >
    <input
      className="delete-button"
      onClick={[Function]}
      type="button"
      value="Удалить"
    />
  </td>
</tr>
`;

exports[`работа кнопки "Редактировать" 2`] = `
<tr
  className="MobileClient"
>
  <td
    className="MobileClientBalance"
  >
    Иванов
  </td>
  <td
    className="MobileClientFIO"
  >
    Иван
  </td>
  <td
    className="MobileClientBalance"
  >
    Иванович
  </td>
  <td
    className="MobileClientFIO"
  >
    200
  </td>
  <td
    className="MobileClientStatusactive"
  >
    active
  </td>
  <td
    className="MobileClientEditButton"
  >
    <input
      className="edit-button"
      disabled={undefined}
      onClick={[Function]}
      type="button"
      value="Редактировать"
    />
  </td>
  <td
    className="MobileClientDeleteButton"
  >
    <input
      className="delete-button"
      onClick={[Function]}
      type="button"
      value="Удалить"
    />
  </td>
</tr>
`;

exports[`работа кнопки "Редактировать" 3`] = `
<tr
  className="MobileClient"
>
  <td
    className="MobileClientBalance"
  >
    Иванов
  </td>
  <td
    className="MobileClientFIO"
  >
    Иван
  </td>
  <td
    className="MobileClientBalance"
  >
    Иванович
  </td>
  <td
    className="MobileClientFIO"
  >
    200
  </td>
  <td
    className="MobileClientStatusactive"
  >
    active
  </td>
  <td
    className="MobileClientEditButton"
  >
    <input
      className="edit-button"
      disabled={undefined}
      onClick={[Function]}
      type="button"
      value="Редактировать"
    />
  </td>
  <td
    className="MobileClientDeleteButton"
  >
    <input
      className="delete-button"
      onClick={[Function]}
      type="button"
      value="Удалить"
    />
  </td>
</tr>
`;

exports[`работа кнопки "Удалить" 1`] = `
<tr
  className="MobileClient"
>
  <td
    className="MobileClientBalance"
  >
    Иванов
  </td>
  <td
    className="MobileClientFIO"
  >
    Иван
  </td>
  <td
    className="MobileClientBalance"
  >
    Иванович
  </td>
  <td
    className="MobileClientFIO"
  >
    200
  </td>
  <td
    className="MobileClientStatusactive"
  >
    active
  </td>
  <td
    className="MobileClientEditButton"
  >
    <input
      className="edit-button"
      disabled={undefined}
      onClick={[Function]}
      type="button"
      value="Редактировать"
    />
  </td>
  <td
    className="MobileClientDeleteButton"
  >
    <input
      className="delete-button"
      onClick={[Function]}
      type="button"
      value="Удалить"
    />
  </td>
</tr>
`;

exports[`работа кнопки "Удалить" 2`] = `
<tr
  className="MobileClient"
>
  <td
    className="MobileClientBalance"
  >
    Иванов
  </td>
  <td
    className="MobileClientFIO"
  >
    Иван
  </td>
  <td
    className="MobileClientBalance"
  >
    Иванович
  </td>
  <td
    className="MobileClientFIO"
  >
    200
  </td>
  <td
    className="MobileClientStatusactive"
  >
    active
  </td>
  <td
    className="MobileClientEditButton"
  >
    <input
      className="edit-button"
      disabled={undefined}
      onClick={[Function]}
      type="button"
      value="Редактировать"
    />
  </td>
  <td
    className="MobileClientDeleteButton"
  >
    <input
      className="delete-button"
      onClick={[Function]}
      type="button"
      value="Удалить"
    />
  </td>
</tr>
`;
